### ISSUE TITLE
Documentation OnsenUI is way below par

### ISSUE THREAD

**Issue by svdoever** [OnsenUI/OnsenUI#1547]
Dear people creating OnsenUI,

I have been doing a lot of mobile development using ReactJS and decided to move over to OnsenUI as my UI framework.Features looked good, performance looked good, documentation looked ok on first inspection.... 

I'm  working with the OnsenUI (in combination with ReactJS) for a few months now, and the thing that worries me most is the bad documentation provided with OnsenUI. A lot of details are missing, there are bugs in the documentation, and the documentation is incomplete.

An example:

https://onsen.io/v2/docs/react/Navigator.html
- No documentation on the objects to be passed to the props, what is the definition of the route object, what object os passed on the on... functions like onPrePush()?
- Prop **animation** is missing, but the documentation of the prop is provided (documentation bug)
- All methods are missing the **route** parameter (documentation bug)
- All methods are missing the documentation of the **options** parameter, instead the text "TypeError: Cannot read property 'toString' of undefined"
- Not all methods are documented if I should believe the Typescript type definition file which reports the following methods:
  
  interface OnsNavigatorElement {
    popPage(options?: navigatorOptions): Promise<HTMLElement>;
    pushPage(page: any, options?: PushPageOptions): Promise<HTMLElement>;
    replacePage(page: any, options?: ReplacePageOptions): Promise<HTMLElement>;
    insertPage(index: number, page: any, options?: navigatorOptions): Promise<HTMLElement>;
    resetToPage(page: any, options?: navigatorOptions): Promise<HTMLElement>;
    bringPageTop(item: any, options?: Object): Promise<HTMLElement>;
  }

And this is just one page of the documentation.

One of the best resources I could find to fill in the blanks is the Typescript type definition file found at /Users **node_modules/onsenui/js/onsenui.d.ts**. Here you can see the definition of the objects to be passed to the different functions.

Please please please improve the documentation of OnsenUI, otherwise it will be another UI framework doomed to fail, and to be abandoned by me.

**Comment by munsterlander** [comment]
@svdoever I agree 100% that documentation is necessary for development and advancement.  I would also like to note, that like you, I have abandoned many products due to overall issues such as, lack of documentation, to early for release, or lack of support.  I say that to argue on behalf of the Onsen team.

Their product is hands down the simplest to implement.  Yes, there are some deficiencies but their team has been amazing at working around the clock to address them.  Also, Onsen 2.0 is still in beta (in release candidate stage actually) and I might add is open source.  I have been trying to help out in the documentation side for Vanilla JS and if I knew more about TypeScript I would help out there.  

With that, I would like to point out these great resources that users can also help out with.

https://community.onsen.io/ - Great forums with lots of activity and information

http://tutorial.onsen.io/ - One area of documentation that users can contribute

https://github.com/OnsenUI - All of the repos where you can change the actual documentation for:

https://onsen.io/v2/docs/js.html

**Comment by svdoever** [comment]
@munsterlander I would like OnsenUI to succeed, that is why I wrote down my concern. Because it is open source it might be best to help with producing pull requests on the documentation (or actually the source code the documentation is produced from.  By the way, thanks for the great answers you are giving on issues, helped me out a few times.



Verzonden vanaf mijn Samsung Galaxy-smartphone.


-------- Oorspronkelijk bericht --------
Van: Munsterlander <notifications@github.com>
Datum: 15-08-16 17:56 (GMT+01:00)
Aan: OnsenUI/OnsenUI <OnsenUI@noreply.github.com>
Cc: Serge van den Oever <Serge.van.den.Oever@macaw.nl>, Mention <mention@noreply.github.com>
Onderwerp: Re: [OnsenUI/OnsenUI] Documentation OnsenUI is way below par (#1547)


@svdoever<https://github.com/svdoever> I agree 100% that documentation is necessary for development and advancement. I would also like to note, that like you, I have abandoned many products due to overall issues such as, lack of documentation, to early for release, or lack of support. I say that to argue on behalf of the Onsen team.

Their product is hands down the simplest to implement. Yes, there are some deficiencies but their team has been amazing at working around the clock to address them. Also, Onsen 2.0 is still in beta (in release candidate stage actually) and I might add is open source. I have been trying to help out in the documentation side for Vanilla JS and if I knew more about TypeScript I would help out there.

With that, I would like to point out these great resources that users can also help out with.

https://community.onsen.io/ - Great forums with lots of activity and information

http://tutorial.onsen.io/ - One area of documentation that users can contribute

https://github.com/OnsenUI - All of the repos where you can change the actual documentation for:

https://onsen.io/v2/docs/js.html

-

**Comment by munsterlander** [comment]
@svdoever Thanks and I as well want them to succeed.  I had posted in the forums all the different platforms that we went through before we got to Onsen and now everyone in my group who is using it, only wants to use it.  We believe that Onsen crushes Ionic, et. al.  Ionic definitely is more established, but I think if we all help out, then Onsen will become the powerhouse on the block.  

Also, glad my answers help.  I just started trying to give back because it was originally @frankdiox , @argelius, @IliaSky and @philolo1 who really helped me in the beginning and I saw people having the same questions as me, so I figured, since I know (I think I do! lol) the answer, lets share the knowledge.  Unfortunately, I am a Vanilla dev and all these new frameworks are foreign to me so when people have questions about Angular, React, Meteor, TypeScript, etc. I have no answers.

**Comment by argelius** [comment]
@svdoever 

Thank you for all the great feedback. We also want to make the documentation better.

It looks like there is a bug on the React documentation. We will take a look at that.

If you want to contribute to the docs you can make a PR to the website repo. Most of the docs are located in this directory:

https://github.com/OnsenUI/onsen.io/tree/master/src/documents_en/v2/docs

Also, the reference is generated from comments in the source code. For the React components it looks like this:

https://github.com/OnsenUI/react-onsenui/blob/master/src/components/AlertDialog.jsx

**Comment by beebase** [comment]
I still have no idea how to push a Component Page into the Navigator route stack. All examples I can find demonstrate navigation in one single mickey mouse file, which is useless. In real life,  pages are always in separate files/components which need to be pushed into the Navigator.  Can someone please link me to any docs on this?

**Comment by argelius** [comment]
@beebase

I am on my phone right now so I can't write a long response.

Basically the implementation of `renderPage` is up to the developer. It takes a `route` object and renders a page.

Here is an example of how it could look:

https://github.com/argelius/react-onsenui-redux-weather/blob/master/components/App.js

In this app no custom props are passed to the pages but you could add something like `route.props` and use the spread operator:

```
const renderPage = (route, navigator) => (
  <route.component key={route.key} ...route.props navigator={navigator} />
);
```

I hope this was helpful! If not please tell me and I will clarify.

**Comment by beebase** [comment]
@argelius: thanks! My first attempt with onsenui/react wast fiddling around with the splitter example, trying to embed stack navigation.(too much at once obviously)
After installing the navigation example everything was instantly clear to me.  It perfectly demonstrates pushing multiple page files to the nav stack.
Thanks again for replying so fast.

**Comment by andrewlorenz** [comment]
It can't escape people's notice that this thread is now 2 years old, and yet I can affirm that the (React) documentation is still very poor.  I have only succeeded in getting my own code to work by much trial & error, and by piecing things together from the document plus much reviewing of code in the sample apps.  I have found much that either isn't documented, poorly documented, or (at worse) inaccurately documented.  I concur with others' opinion that this otherwise brilliant project is really let down by the documentation, and counts significantly against it in terms of committing to using it in a serious project.  However, I also understand that this isn't necessarily a priority for the project team, and that the community could help by submitting PR's.  I am going to endeavour to do so, and encourage others to do the same.
